index.js

```diff
@@ -82,6 +87,84 @@ const sendResponse = ( req, settings, fn ) => {
+	// the last line, which represents the whole response in non-$stream mode, we need to convey the
+	// HTTP status code after the body has started. For the other lines, we need an unambiguous way
+	// to know that they’re not the last line, so we can exclude it without a “delay line”.
+	if ( expectStreamMode ) {
+		// Streaming responses is trickier than you might expect, with many footguns:
+		// • req.buffer(false): no version of superagent implements this when running in the browser
+		// • req.parse() or superagent.parse[]: only gets called when the response ends (see above)
+		// • req.on("progress"): doesn’t seem to work... at all
+		// • req.responseType(anything): makes superagent skip parse functions (see above)
+		// • req.xhr.responseType="blob": XHR only exposes partial responses in "" or "text" modes
+		// • req.xhr: only available after you call req.end()
+
+		// Expose partial responses.
+		// <https://xhr.spec.whatwg.org/#the-response-attribute>
+		req.xhr.responseType = 'text';
+
+		// Force the response to be treated as UTF-8. This is easier than the x-user-defined trick
+		// that would otherwise be needed to access the raw binary response body.
+		// <https://xhr.spec.whatwg.org/#final-charset>
+		// <https://github.com/ndjson/ndjson-spec/blob/1.0/README.md#31-serialization>
+		// <https://stackoverflow.com/a/33042003>
+		req.xhr.overrideMimeType( 'application/x-ndjson; charset=utf-8' );
+
+		// Find response chunks that end in a newline (possibly preceded by a carriage return), then
+		// for each chunk except the last, parse it as JSON and pass that to onStreamRecord.
+		// <https://github.com/ndjson/ndjson-spec/blob/1.0/README.md#31-serialization>
+		// <https://github.com/ndjson/ndjson-spec/blob/1.0/README.md#32-parsing>
+		// <https://stackoverflow.com/a/38440028>
+		let lastLine = null;
+		let start = 0;
+
+		req.xhr.addEventListener( 'progress', ( { target } ) => {
+			// Don’t use ProgressEvent#loaded in this algorithm. It measures progress in octets,
+			// while we’re working with text that has already been decoded from UTF-8 into a string
+			// that can only be indexed in UTF-16 code units. Reconciling this difference is not
+			// worth the effort, and might even be impossible if there were encoding errors.
+			let stop;
+			while ( true ) {
+				stop = target.response.indexOf( '\n', start );
+
+				if ( stop < 0 ) {
+					// Leave start untouched for the next progress event, waiting for the newline
+					// that indicates we’ve finished receiving a full line.
+					break;
+				}
+
+				lastLine = target.response.slice( start, stop );
+
+				// Note: not ignoring empty lines.
+				// <https://github.com/ndjson/ndjson-spec/blob/1.0/README.md#32-parsing>
+				const record = JSON.parse( lastLine.trim() );
+
+				// Non-last lines should have .status == 100.
+				if ( record.status < 200 ) {
+					debug( 'stream mode: record=%o', record );
+					onStreamRecord( record.body );
```

Comment:
An alternative implementation could fire events on the XHR object:
```js
const event = new Event( 'streamrecord' );
event.record = record.body;
req.xhr.dispatchEvent( event );
```
and used as:
```js
const req = wpcom.req( ..., ( err, response ) => { ... } );
req.addEventListener( 'streamrecord', ( e ) => {
  console.log( e.record );
} );
```
I'm not sure if it's better or worse, just pointing out it's possible 🙂 In the same way, you would listen for `progress` events on the request.

Reply:
interesting idea! i would personally lean towards staying out of XHR’s event namespace, in case the platform adds a new event with the same name someday, but i would be very surprised if that ever happens.